File: exercises/ddd_application_layer/typescript/src/domain/application/Application.ts
import type { ApplicationId } from "./ApplicationId";
import type { ApplicationReceivedDateTime } from "./ApplicationReceivedDateTime";
import type { EmailAddress } from "./EmailAddress";
import type { FullName } from "./FullName";
import type { MotivationLetter } from "./MotivationLetter";
import type { PhoneNumber } from "./PhoneNumber";

export class Application {
  private _id: ApplicationId;
  private _receivedDateTime: ApplicationReceivedDateTime;
  private _fullName?: FullName;
  private _emailAddress?: EmailAddress;
  private _phoneNumber?: PhoneNumber;
  private _motivationLetter?: MotivationLetter;

  get id() { return this._id; }
  get receivedDateTime() { return this._receivedDateTime; }
  get fullName() { return this._fullName; }
  get emailAddress() { return this._emailAddress; }
  get phoneNumber() { return this._phoneNumber; }
  get motivationLetter() { return this._motivationLetter; }

  constructor(id: ApplicationId, receivedDateTime: ApplicationReceivedDateTime) {
    if (!id || !receivedDateTime) {
      throw new Error("id and receivedDateTime are required");
    }
    this._id = id;
    this._receivedDateTime = receivedDateTime;
  }

  fillOutBasicDetails(fullName: FullName, emailAddress: EmailAddress, phoneNumber: PhoneNumber) {
    this._fullName = fullName;
    this._emailAddress = emailAddress;
    this._phoneNumber = phoneNumber;
  }

  submitMotivationLetter(motivationLetter: MotivationLetter) {
    if (!this.areBasicDetailsFilledOut()) {
      throw new Error("Cannot accept motivation letter before basic details are filled out");
    }
    this._motivationLetter = motivationLetter;
  }

  equals(other: Application) {
    return this._id.equals(other._id);
  }

  private areBasicDetailsFilledOut(): boolean {
    return !!(this._fullName && this._emailAddress && this._phoneNumber);
  }
}

File: exercises/ddd_application_layer/java/src/main/java/domain/application/Application.java
package domain.application;
import java.util.Objects;

public class Application {
    private ApplicationId id;
    private ApplicationReceivedDateTime receivedDateTime;
    private FullName fullName;
    private EmailAddress emailAddress;
    private PhoneNumber phoneNumber;
    private MotivationLetter motivationLetter;

    public Application(ApplicationId id, ApplicationReceivedDateTime receivedDateTime) {
        Objects.requireNonNull(id, "ApplicationId cannot be null");
        Objects.requireNonNull(receivedDateTime, "ApplicationReceivedDateTime cannot be null");
        this.id = id;
        this.receivedDateTime = receivedDateTime;
    }

    public ApplicationId getId() { return id; }
    public ApplicationReceivedDateTime getReceivedDateTime() { return receivedDateTime; }
    public FullName getFullName() { return fullName; }
    public EmailAddress getEmailAddress() { return emailAddress; }
    public PhoneNumber getPhoneNumber() { return phoneNumber; }
    public MotivationLetter getMotivationLetter() { return motivationLetter; }

    public void fillOutBasicDetails(FullName fullName, EmailAddress emailAddress, PhoneNumber phoneNumber) {
        Objects.requireNonNull(fullName, "FullName cannot be null");
        Objects.requireNonNull(emailAddress, "EmailAddress cannot be null");
        Objects.requireNonNull(phoneNumber, "PhoneNumber cannot be null");
        this.fullName = fullName;
        this.emailAddress = emailAddress;
        this.phoneNumber = phoneNumber;
    }

    public void fillOutMotivationLetter(MotivationLetter letter) {
        Objects.requireNonNull(letter, "MotivationLetter cannot be null");
        if (!areBasicDetailsFilledOut()) {
            throw new IllegalStateException("Basic details must be filled out before adding a motivation letter.");
        }
        this.motivationLetter = letter;
    }

    public boolean equals(Application other) {
        return this.id.equals(other.id);
    }

    private boolean areBasicDetailsFilledOut() {
        return fullName != null && emailAddress != null && phoneNumber != null;
    }
}

